syntax = "proto3";

package kv_cache_manager.proto.debug;

option cc_generic_services = false;

enum ErrorCode {
    UNSPECIFIED = 0;
    OK = 1;
    UNSUPPORTED = 2;
    INTERNAL_ERROR = 3;
    SERVICE_NOT_READY = 4;
    INVALID_ARGUMENT = 5;

    IO_ERROR = 20;
    UNKNOWN_ERROR = 100;
    ERROR_MAX = 65535;
}

message Status {
    ErrorCode code = 1;
    string message = 2;
}

message CommonResponseHeader {
    Status status = 1;
    string request_id = 2;
}

message CommonResponse { CommonResponseHeader header = 1; }


enum FaultType {
    FT_INTERNAL_ERROR = 0;
}

enum FaultTriggerStrategy {
    ALWAYS = 0; // 总是注入故障
    ONCE = 1; // 在某次请求时注入故障
}

message InjectFaultRequest {
    string api_name = 1;  
    FaultType fault_type = 2;
    FaultTriggerStrategy fault_trigger_strategy= 3;
    int32 trigger_at_call = 4; // FaultTriggerStrategy::ONCE时某个方法第几次被调用时注入故障
}

message RemoveFaultRequest {
  string api_name = 1;  
}

message ClearFaultsRequest {}

service DebugService { 
    // 给某个方法注入故障
    rpc InjectFault(InjectFaultRequest) returns (CommonResponse);
    // 给某个方法清除故障
    rpc RemoveFault(RemoveFaultRequest) returns (CommonResponse);
    // 清除所有方法故障
    rpc ClearFaults(ClearFaultsRequest) returns (CommonResponse);
}